library/modules/EventManager.cpp

```diff
@@ -1254,3 +1261,32 @@ static void manageInteractionEvent(color_ostream& out) {
         //TODO: deduce attacker from latest defend event first
     }
 }
+
+static void manageActionEvent(color_ostream& out) {
+    if (!df::global::world)
+        return;
+    multimap<Plugin*,EventHandler> copy(handlers[EventType::UNIT_ACTION].begin(), handlers[EventType::UNIT_ACTION].end());
+    for ( size_t a = 0; a < df::global::world->units.all.size(); a++ ) {
+        df::unit* unit = df::global::world->units.all[a];
+        if ( Units::isActive(unit) ) {
+            auto knownActions = &unitToKnownActions[unit->id];
+            for ( df::unit_action* action : unit->actions ) {
+                if ( action->type != df::unit_action_type::None) {
+                    if ( std::find(knownActions->begin(), knownActions->end(), action->id) == knownActions->end() ) {
+                        knownActions->push_back(action->id);
+                        for ( auto b = copy.begin(); b != copy.end(); b++ ) {
+                            EventHandler handle = (*b).second;
+                            ActionData data = {unit->id, action, action->id};
+                            handle.eventHandler(out, (void*)&data);
+                        }
+                    }
+                } else {
+                    auto newEnd = std::remove(knownActions->begin(), knownActions->end(), action->id);
+                    knownActions->erase(newEnd, knownActions->end());
```

Comment:
Could you explain what is happening here? Is an action type of None invalidating all more recent actions?

Reply:
It's the [erase-removed idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom), though perhaps more verbose than it needs to be

Reply:
I see. So it's only removing a single element. Does order matter for these actions, or would we be better off using a set instead of a vector?